accounts, cmd, eth, internal, mobile, node: split account backends

karalabe opened the pull request:
# This PR

This PR does the first step towards the proposal described below. It splits up the current account management into a top level "aggregation layer", and lower level account backends. The previous AccountManager was renamed to KeyStore (since that's what it actually was) and moved into its own backend.

This PR does not contain either hardware extensions nor remote signers yet, only the refactors necessary to make those possible. Similarly the top level caching layer still needs some Go API level events to be put in place, so currently there's no top cache (the existing file cache for keystores remains in place, so there's no performance loss compared to the old code).

## Caveats

@fjl proposed that we should not pass the `chainID` for every signing operation, rather pass it only once when we initialize the backend. Unfortunately this cannot be done with the current code as it signs depending on whether we progressed over EIP155 or not. Whether we want to change this is debatable, just saying the reason. The chain ID is also unknown at keystore initialization as it's read from the database's chain config, so it either needs to wait for eth to be initialized, or needs to peek into the database yet again.

# Background

Currently our accounts package is quite simplistic: it features an ecrypted keystore hard coded into the account manager, and all transaction signing goes through it; and we also have an address cache in place to cache available accounts as well as to detect newly added/removed addresses via file system notifications.

The problem with the current design is that it's not easily extendable to support alternative account storage and/or transaction signing mechanisms such as hardware wallets or alternative schemes. This proposal aims to rework the accounts package a bit so that we can add new authorization mechanisms.

## Account backends

To support alternative authorization mechanisms, we need to introduce the concept of account backends, whereby we abstract out the high level operations we want to perform with accounts (e.g. sign transaction) from the low level operations specific to some mechanism (e.g. decrypt [web3 ssd](https://github.com/ethereum/wiki/wiki/Web3-Secret-Storage-Definition) key, vs. retrieve ledger feature set). Similarly we need to delegate watching for account changes to these backends and only handle the fired events in the high level account manager.

First, we modify the `accounts.Account` structure a bit. Currently it contains an `Address` field for identifying an account by its Ethereum address, and an optional `File` field so that in case of duplicate or external file-system accounts, the correct one can be decrypted. Since the name `File` doesn't really make sense in the scope of a generic backend, it should be renamed to `URL`.

```go
type Account struct {
	Address common.Address // Ethereum account address derived from the key
	URL     string         // Optional resource locator within a backend
}
```

We also need to define the backend interface that all implementations such as a password protected keystore or a hardware wallet would need to implement. The important thing to note is that we cannot only send transaction hashes to sign any more, since some account backends would need the transactions details too to do user confirmations before signing, so the new API needs to work with transactions. This also entails that individual backends need to be made aware of chain IDs, as they are the ones generating the hash to sign (we could pass the chainID as the V parameter of the transaction, but that would make a very subtle API that people could too easily misuse). 

```go

// Backend is an "account provider" that can specify a batch of accounts it can
// sign transactions with and upon request, do so.
type Backend interface {
	// Accounts retrieves the list of signing accounts the backend is currently aware of.
	Accounts() []Account

	// HasAddress reports whether an account with the given address is present.
	HasAddress(addr common.Address) bool

	// SignHash requests the backend to sign the given hash.
	//
	// It looks up the account specified either solely via its address contained within,
	// or optionally with the aid of any location metadata from the embedded URL field.
	SignHash(acc Account, hash []byte) ([]byte, error)

	// SignTx requests the backend to sign the given transaction.
	//
	// It looks up the account specified either solely via its address contained within,
	// or optionally with the aid of any location metadata from the embedded URL field.
	//
	// If the backend does not support EIP155 based chain IDs, it will sign using the
	// original homestead signing rules, making the transaction valid on all Ethereum
	// chains.
	SignTx(acc Account, tx *types.Transaction, chainID *big.Int) (*types.Transaction, error)

	// SignHashWithPassphrase requests the backend to sign the given transaction with
	// the given passphrase as extra authentication information.
	//
	// It looks up the account specified either solely via its address contained within,
	// or optionally with the aid of any location metadata from the embedded URL field.
	SignHashWithPassphrase(acc Account, passphrase string, hash []byte) ([]byte, error)

	// SignTxWithPassphrase requests the backend to sign the given transaction, with the
	// given passphrase as extra authentication information.
	//
	// It looks up the account specified either solely via its address contained within,
	// or optionally with the aid of any location metadata from the embedded URL field.
	//
	// If the backend does not support EIP155 based chain IDs, it will sign using the
	// original homestead signing rules, making the transaction valid on all Ethereum
	// chains.
	SignTxWithPassphrase(acc Account, passphrase string, tx *types.Transaction, chainID *big.Int) (*types.Transaction, error)

	// Watch requests the backend to send a notification to the specified channel whenever
	// an new account appears or an existing one disappears.
	Watch(chan AccountEvent) error

	// Unwatch requests the backend stop sending notifications to the given channel.
	Unwatch(chan AccountEvent) error
}
```

An already mentioned, the backend needs to be able to monitor for account changes (e.g. new account copied into encrypted keystore folder or USB hardware wallet plugged into the system). These events can be subscribed to through the backend's `Watch` and `Unwatch` methods, which will send `AccountEvent` notifications.

```go
type AccountEvent struct {
	Account Account
	Added   bool
}
```

### Keystore backend

With the current separation proposal, the keystore needs to be split out of the account manager as currently the two are implicitly the same. The biggest impact this has on the code is that all the methods that operate on keystores (account creation, deletion, update, etc) need to move out of the account manager too into the keystore backend, as they don't make any sense in the context of other backends. This inherently means, that Geth will need to be able to obtain a reference to the keystore itself too, not just the account manager.

A minor cimplication in the refactor will be that the various `Signer` handling will need to be done inside the `Keystore` backend from now (as they rely on private key availability) and not other parts of the code.

### Hardware backend

With the backends separated per the above, hardware wallets can be fairly easily added. One example of supporting the Ledger Nano S would be a backend that only implements transaction signing (it cannot sign data as of now, so those methods would return an error). Account listing and watching could simply be implemented via USB scanning and hotplug monitoring.

This approach does mean that we have to implement the wallet protocol for every wallet we want to support ourselves (or at least convince someone to implement it in Go for us). This is a debated topic, but in my opinion having out of the box support for at least the mayor players would be welcome by everyone in the community and I prefer pragratic tools vs *pure* ones that are hard to use.

### External backend

Beside the common use cases and hardware wallets, it may easily happen that someone comes up with a fancy new way to sign transactions. It should be possible for them to do so without any explicit modification to the `go-ethereum` library. This however entails that external code needs to be executed upon request.

One such approach would be a plugin infrastructure. Unfortunately there is so simple way to do such thing in Go. The 1.8 release does have some experimental support for loading plugins via shared libraries, but it only works on Linux, and it needs the library to be written in Go, which isn't that common yet in the crypto world. The other popular languages that people could use would all require embedded interpreters or compilers, so they are a whole new can of worms.

The alternative proposal is an RPC API based remote signer. This would entail creating an accounts backend which is actually an RPC service, providing its own `signer` API namespace. As such, any locally running program could attach to one ouf our API interfaces (IPC or WebSocket, as it needs push support), and register accounts bound to the lifetime of the connection. Then whenever the backend receives a request to sign a transaction, that is pushed in JSON RPC format to one of the attached signer clients, which may reply with a signature or an error.

The API of the external signer would look like:

 * `signer_register({address, url})` [Signer->Geth]: Make a new account available in the account manager
 * `signer_unregister({address, url})` [Signer->Geth]: Make a new account unavailable in the account manager
 * `signer_signTx({address, url}, tx, chainid, [passphrase])` [Geth->Signer]: Request signing a transaction
   * Signer returns either a signed transaction or an error
 * `signer_signHash({address, url}, hash, [passphrase])` [Geth->Signer]: Request signing an arbitrary data hash
    * Signer returns either a signature blob or an error

*Although this approach looks promising and in theory covers all signing requirements, it does require third party vendors to provide proper signers for their devices/services that run on all needed platforms. As Ethereum is quite new to the scene, and Geth is but one of many clients, expecting major vendors to so their share of the work is imho naive.*

## Account manager and address cache

Given the account listing property of the `Backend` interface and it's functionality to watch for changes, the address cache remains simple to implement (essentially the current one with the keystore monitoring plucked out).

Regarding the account manager, it should gain a couple methods to register and drop backends on demand (to support writing custom backend in Go when integrating `go-ethereum` into other projects).

Arachnid commented on accounts/accounts.go:
```diff
@@ -0,0 +1,178 @@
+// Copyright 2015 The go-ethereum Authors
```
2017

karalabe replied:
Well, this file started out in 2015, just that it was split, so git figured one file is moved while the other is new. I think it should remain 2015 though as that's when it started out.

Arachnid replied:
If there's significant changes, the copyright ought to be updated. Is it just a split, or did you change the APIs?

karalabe replied:
Fixed.

Arachnid commented on accounts/accounts.go:
```diff
@@ -0,0 +1,178 @@
+	return []byte(`"` + acc.Address.Hex() + `"`), nil
+}
+
+func (acc *Account) UnmarshalJSON(raw []byte) error {
```
Is this functional?

karalabe replied:
Good question. @fjl?

Arachnid commented on accounts/accounts.go:
```diff
@@ -0,0 +1,178 @@
+// Manager is an overarching account manager that can communicate with various
+// backends for signing transactions.
+type Manager struct {
+	backends []Backend                // List of currently registered backends (ordered)
```
ordered by what? And do we need the list as well as the map?

karalabe replied:
By registration order. Since someone "invented" the concept of "primary account" and account unlocking by index, unfortunately we need to hae a strict order on the accounts, hence why the slice is needed.

As for the map, it's needed to allow quick lookups for backends without needing to scan over the previous list.

Arachnid replied:
Could you mention in the comment what it's ordered by?

Arachnid commented on accounts/accounts.go:
```diff
@@ -0,0 +1,178 @@
+
+// SignTx requests the account manager to get the transaction signed with an
+// arbitrary signing backend holding the authorization for the specified account.
+func (am *Manager) SignTx(acc Account, tx *types.Transaction, chainID *big.Int) (*types.Transaction, error) {
```
Does chainID need to be a `*big.Int`?

karalabe replied:
As far as I know, chain IDs are defined to be 256bits in size.

Arachnid replied:
I thought they were 0-255? They're added to the 1-byte V value, aren't they?

karalabe replied:
Not sure, @obscuren ?

obscuren replied:
Chain IDs are 256 bit.

Arachnid commented on accounts/accounts.go:
```diff
@@ -0,0 +1,178 @@
+
+// SignHashWithPassphrase requests the account manager to get the hash signed with
+// an arbitrary signing backend holding the authorization for the specified account.
+func (am *Manager) SignHashWithPassphrase(acc Account, passphrase string, hash []byte) ([]byte, error) {
```
How do callers know when to call `*WithPassPhrase` and when to call the regular methods?

karalabe replied:
This is really a way to specify an extra field of information to the backend for signing which it either uses or not (hw wallets won't use it, keystores will). I couldn't figure out a cleaner way to support this operation without requiring user input callbacks.

Arachnid replied:
From the POV of a user, though, there's no clear way to know which method to call if you don't know anything about the backend for an account. Could we specify that the regular sign methods will return a specific error if a password is required?

karalabe replied:
I've added an `AuthNeededError` that should be returned if the user needs to call the other function (or execute some other action such as unlock the account).

Arachnid commented on accounts/backend.go:
```diff
@@ -0,0 +1,82 @@
+// Backend is an "account provider" that can specify a batch of accounts it can
+// sign transactions with and upon request, do so.
+type Backend interface {
+	// Accounts retrieves the list of signing accounts the backend is currently aware of.
```
It's worth considering how we should handle this for HDR wallets.

karalabe replied:
The same way we do it for keystores. The backend maintains a local cache of accounts it's currently aware of and returns that. The keystore registeres filesystem watches and updates its cache on change (or periodically on android). The same can be done for HW wallets, where the backend monitors USB events, and maintains a list of supported devices.

Arachnid replied:
In that event, though, we should at least document that the list of accounts returned may be comprehensive; eg, that it's okay for `HasAccount` to return true for an account not in the list.

karalabe replied:
Ah, I though you meant hardware wallets. I'm not sure how to do HD wallets currently.

Arachnid commented on accounts/accounts.go:
```diff
@@ -0,0 +1,178 @@
+// Account represents a stored key.
+// When used as an argument, it selects a unique key to act on.
+type Account struct {
+	Address common.Address // Ethereum account address derived from the key
```
Do we need both address and URL? Could we ask the backend for the address for a given account when we need it?

karalabe replied:
The point of the URL is that an address may have multiple backing keys in the case of the keystore. So the reason this fields was introduced is to allow differentiating between them.

karalabe replied:
It also helps if for example you want to sign something with a key that's actually outside of the keystore. Then this field can be used to read the key data.

Arachnid replied:
Not sure I follow; if you had URL but not address, couldn't you look up whatever you needed in the relevant backend?

karalabe replied:
Not currently.

karalabe replied:
We might implement something towards that end, just saying there's no such logic currently, and I'd rather not expand the scope of this PR.

Arachnid replied:
Would it be hard to add, though? For the accounts backend, the URI would presumably just be something like the name of the file containing the key.

Arachnid commented on accounts/backends/keystore/keystore.go:
```diff
@@ -0,0 +1,362 @@
+// Copyright 2015 The go-ethereum Authors
```
2017

karalabe replied:
Same, this is essentially the AccountManager, renamed.

Arachnid commented on mobile/keystore.go:
```diff
@@ -79,57 +80,73 @@ func (a *Account) GetAddress() *Address {
 // [R || S || V] format where V is 0 or 1.
-func (am *AccountManager) SignPassphrase(account *Account, passphrase string, hash []byte) (signature []byte, _ error) {
-	return am.manager.SignWithPassphrase(account.account, passphrase, hash)
+func (ks *KeyStore) SignHashPassphrase(account *Account, passphrase string, hash []byte) (signature []byte, _ error) {
```
Shouldn't this be `SignHashWithPassphrase`?

karalabe replied:
Sadly Swift 3 screws up function renaming, and writing "WithSomething" will cut the suffix off, and rename the first parameter to "Something", also adding a requirement to explicitly specify it. So it would look

`swiftKeyStore.signHash(passphrase: myAccount, ...)`

Arachnid replied:
So, is this not intended to be an implementation of the backend interface? Could we document as much somewhere?

karalabe replied:
This is just the mobile API. It's a simplified version of what the generally support in go-ethereum.

Arachnid commented on mobile/keystore.go:
```diff
@@ -79,57 +80,73 @@ func (a *Account) GetAddress() *Address {
+
+// SignTxPassphrase signs the transaction if the private key matching the
+// given address can be decrypted with the given passphrase.
+func (ks *KeyStore) SignTxPassphrase(account *Account, passphrase string, tx *Transaction, chainID *BigInt) (*Transaction, error) {
```
Shouldn't this be `SignTxWithPhassphrase`?

karalabe replied:
Please see above :)

karalabe commented:
@Arachnid PTAL.

@fjl I've moved the `keystore` one package nesting out (so we don't have a `backends` package any more). Should look nicer, though I still maintain that having a separate package for it makes it easier to both reuse and also ensures it's not "special" :P

I've also introduced 

```go
// AuthNeededError is returned by backends for signing requests where the user
// is required to provide further authentication before signing can succeed.
//
// This usually means either that a password needs to be supplied, or perhaps a
// one time PIN code displayed by some hardware device.
type AuthNeededError struct {
	Needed string // Extra authentication the user needs to provide
}
```

which for example the kestore uses as 

```ErrNeedPasswordOrUnlock = accounts.NewAuthNeededError("password or unlock")```

in place of the old `ErrLocked` error.

karalabe commented:
I am closing this PR in favor of https://github.com/ethereum/go-ethereum/pull/3592. Since we further designed how the accounts should work, I don't want to reimplement the design both in this and the Ledger PR. I'll keep working on that thread.
